src/vos/vos_dtx.c

```diff
@@ -717,6 +730,469 @@ vos_dtx_abort_one(struct vos_container *cont, struct daos_tx_id *dti,
+					  "Invalid shared key DTX shares %d\n",
+					  key->kr_dtx_shares);
+				key->kr_dtx_shares = 1;
+				goto add;
+			}
+
+			key->kr_dtx_shares++;
+			shared = true;
+
+			/* The original key DTX has been aborted, but some
+			 * others still share the key. Then set the kr_dtx
+			 * to current DTX.
+			 */
+			if (umem_id_equal(umm, key->kr_dtx,
+					  DTX_UMMID_UNKNOWN)) {
+				key->kr_dtx = dth->dth_ent;
+
+				D_DEBUG(DB_TRACE, "The DTX "DF_DTI" shares key "
+					"with unknown DTXs, shares count %u.\n",
+					DP_DTI(&dth->dth_xid),
+					key->kr_dtx_shares);
+				goto add;
+			}
+
+			sh_dtx = umem_id2ptr(umm, key->kr_dtx);
+			D_ASSERT(dtx != sh_dtx);
+			D_ASSERTF(sh_dtx->te_state == DTX_ST_PREPARED ||
+				  sh_dtx->te_state == DTX_ST_INIT,
+				  "Invalid shared key DTX state: %u\n",
+				  sh_dtx->te_state);
+
+			umem_tx_add_ptr(umm, sh_dtx, sizeof(*sh_dtx));
+			sh_dtx->te_flags |= DTX_EF_SHARES;
+
+			D_DEBUG(DB_TRACE, "The DTX "DF_DTI" shares key with "
+				DF_DTI", the shares count %u.\n",
+				DP_DTI(&dth->dth_xid), DP_DTI(&sh_dtx->te_xid),
+				key->kr_dtx_shares);
+		}
+
+add:
+		rec_mmid = umem_zalloc(umm, sizeof(struct vos_dtx_record_df));
+		if (UMMID_IS_NULL(rec_mmid))
+			return -DER_NOMEM;
+
+		rec = umem_id2ptr(umm, rec_mmid);
+		rec->tr_type = vds->vds_type;
+		rec->tr_flags = 0;
+		rec->tr_record = vds->vds_record;
+
+		rec->tr_next = dtx->te_records;
+		dtx->te_records = rec_mmid;
+
+drop:
+		d_list_del(&vds->vds_link);
+		D_FREE_PTR(vds);
+	}
+
+	if (rc == 0 && shared)
+		dtx->te_flags |= DTX_EF_SHARES;
+
+	return rc;
+}
+
+/* The caller has started PMDK transaction. */
+void
+vos_dtx_degister_record(struct umem_instance *umm,
+			umem_id_t entry, umem_id_t record, uint32_t type)
+{
+	struct vos_dtx_entry_df		*dtx;
+	struct vos_dtx_record_df	*rec;
+	struct vos_dtx_record_df	*prev = NULL;
+	umem_id_t			 rec_mmid;
+
+	if (!vos_dtx_is_normal_entry(umm, entry))
+		return;
+
+	dtx = umem_id2ptr(umm, entry);
+	rec_mmid = dtx->te_records;
+	while (!UMMID_IS_NULL(rec_mmid)) {
+		rec = umem_id2ptr(umm, rec_mmid);
+		if (!umem_id_equal(umm, record, rec->tr_record)) {
+			prev = rec;
+			rec_mmid = rec->tr_next;
+			continue;
+		}
+
+		if (prev == NULL) {
+			umem_tx_add_ptr(umm, dtx, sizeof(*dtx));
+			dtx->te_records = rec->tr_next;
+		} else {
+			umem_tx_add_ptr(umm, prev, sizeof(*prev));
+			prev->tr_next = rec->tr_next;
+		}
+
+		umem_tx_add_ptr(umm, rec, sizeof(*rec));
+		umem_free(umm, rec_mmid);
+		break;
+	};
+
+	/* The caller will destroy related OBJ/KEY/SVT/EVT record after
+	 * degistered the DTX record. So not reset DTX reference inside
+	 * related OBJ/KEY/SVT/EVT record unless necessary.
+	 */
+	if (type == DTX_RT_OBJ) {
+		struct vos_obj_df	*obj;
+
+		obj = umem_id2ptr(umm, record);
+		if (dtx->te_intent == DAOS_INTENT_UPDATE)
+			obj->vo_dtx_shares--;
+		if (umem_id_equal(umm, obj->vo_dtx, entry) &&
+		    obj->vo_dtx_shares > 0) {
+			umem_tx_add_ptr(umm, obj, sizeof(*obj));
```

Comment:
this has to be called before you change obj->vo_dtx_shares

Reply:
If the shares become zero, then the object will be freed.

Reply:
what if it's not zero and you've changed it w/o adding the undo log, then the node crashed? This change will not be rolled back, but everything else will be?

Reply:
So that means if I want to modify something that belongs to the object to be freed, I still need to call umem_tx_add_ptr() for the case of rollback, right?

Reply:
yes, if you changed anything, please call tx_add, and if you only need to free a data structure, then you don't need to tx_add

Reply:
The patch has been refreshed as you suggested.